invokeai/app/api/routers/model_records.py

```diff
@@ -162,3 +171,151 @@ async def add_model_record(
+        "access_token": "f5820a918aaf01"
+      }`
+    The `access_token` field is optonal
+
+    The model's configuration record will be probed and filled in
+    automatically.  To override the default guesses, pass "metadata"
+    with a Dict containing the attributes you wish to override.
+
+    Installation occurs in the background. Either use list_model_install_jobs()
+    to poll for completion, or listen on the event bus for the following events:
+
+      "model_install_started"
+      "model_install_completed"
+      "model_install_error"
+
+    On successful completion, the event's payload will contain the field "key"
+    containing the installed ID of the model. On an error, the event's payload
+    will contain the fields "error_type" and "error" describing the nature of the
+    error and its traceback, respectively.
+
+    """
+    logger = ApiDependencies.invoker.services.logger
+
+    try:
+        installer = ApiDependencies.invoker.services.model_install
+        result: ModelInstallJob = installer.import_model(
+            source=source,
+            config=config,
+        )
+        logger.info(f"Started installation of {source}")
+    except UnknownModelException as e:
+        logger.error(str(e))
+        raise HTTPException(status_code=404, detail=str(e))
+    except InvalidModelException as e:
+        logger.error(str(e))
+        raise HTTPException(status_code=415)
+    except ValueError as e:
+        logger.error(str(e))
+        raise HTTPException(status_code=409, detail=str(e))
+    return result
+
+
+@model_records_router.get(
+    "/import",
+    operation_id="list_model_install_jobs",
+)
+async def list_model_install_jobs(
+    source: Optional[str] = Query(
+        description="Filter list by install source, partial string match.",
+        default=None,
+    ),
+) -> List[ModelInstallJob]:
+    """
+    Return list of model install jobs.
+
+    If the optional 'source' argument is provided, then the list will be filtered
+    for partial string matches against the install source.
+    """
+    jobs: List[ModelInstallJob] = ApiDependencies.invoker.services.model_install.list_jobs(source)
+    return jobs
+
+
+@model_records_router.patch(
+    "/import",
+    operation_id="prune_model_install_jobs",
+    responses={
+        204: {"description": "All completed and errored jobs have been pruned"},
+        400: {"description": "Bad request"},
+    },
+)
+async def prune_model_install_jobs() -> Response:
+    """
+    Prune all completed and errored jobs from the install job list.
+    """
+    ApiDependencies.invoker.services.model_install.prune_jobs()
+    return Response(status_code=204)
+
+
+@model_records_router.patch(
+    "/sync",
+    operation_id="sync_models_to_config",
+    responses={
+        204: {"description": "Model config record database resynced with files on disk"},
+        400: {"description": "Bad request"},
+    },
+)
+async def sync_models_to_config() -> Response:
```

Comment:
The model manager services are relatively well contained -- they don't interact with other services aside from the event bus. Should I try switching them from `async def` to `def` and see what happens?

How does python's asyncio interact with threads? Should I have implemented the multithreaded download queue using asyncio tasks rather than threads?

Reply:
I put some more thought into this. We're definitely not going to be able to solve everything in this PR, but we should start thinking about the direction we want to move in.

These are the main issues that I currently see across the app:
- We do blocking synchronous operations (e.g. DB accesses) in some `async def` routes.
- Many of the 'services' are not rigorously designed to be thread-safe. This means that they cannot safely be used in `def` routes.

As a general direction, I think we should plan to have to support many `def` routes, since lots of third-party libraries (most DB libraries) don't offer an async API. To support this, we will need to start thinking more about whether our services are thread-safe.

For this PR, I'm fine with leaving as `async def` to minimize changes needed. There are so many instances like this already that it's not really adding much more future work.

Regarding the download queue implementation: I haven't looked at that PR yet. I can think of a number of ways to implement it. Did you consider using [FastAPI Background Tasks](https://fastapi.tiangolo.com/tutorial/background-tasks/#background-tasks)? They abstract away most of the threading/async complexity.

Reply:
From https://fastapi.tiangolo.com/tutorial/background-tasks/#background-tasks: 
> It can be an async def or normal def function, FastAPI will know how to handle it correctly.

I interpreted this to mean that it will be run in a thread if defined as a normal `def` function. We can save this discussion for the other PR though.

Reply:
Ok, I'm resolving this conversation.